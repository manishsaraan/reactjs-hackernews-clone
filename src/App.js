import React, { Component } from 'react';
import './App.css';

const DEFAULT_QUERY = "redux";
const PATH_BASE = 'https://hn.algolia.com/api/v1';
const PATH_SEARCH = '/search';
const PARAM_SEARCH = '?query=';
const DEFAULT_PAGE = 0;
const PARAM_PAGE = 'page=';
const DEFAULT_HPP = 25;
const PARAM_HPP = 'hitsPerPage=';

const isSearched = query => item => !query || item.title.toLowerCase().indexOf(query.toLowerCase()) !== -1;

class App extends Component {
  constructor(props){
    super(props);
    this.state = {
      results : null,
      query : DEFAULT_QUERY,
      isAjax : false,
      searchKey : ''
    };
    this.onSearchChange = this.onSearchChange.bind(this);
    this.setSearchTopStories = this.setSearchTopStories.bind(this);
    this.fetchSearchTopStories = this.fetchSearchTopStories.bind(this);
    this.onSubmit = this.onSubmit.bind(this);
    this.needsToSearchTopStories = this.needsToSearchTopStories.bind(this);
  }
  
  //check if data is present is cache or not
  needsToSearchTopStories(query){
    return !this.state.results[query];
  }

  //update state with results from api
  setSearchTopStories(result){
    const {hits, page }  = result;
    const {query, searchKey} = this.state;
    const oldHits  = page ===0 ? [] : this.state.results[searchKey].hits;
    const updateHits = [...oldHits, ...hits];

    this.setState({
      results : { ...this.state.results, [searchKey] : {hits : updateHits, page}}
    });
  }

  //fetch data
  fetchSearchTopStories(query, page){
    let searchQuery = `${PATH_BASE}${PATH_SEARCH}${PARAM_SEARCH}${query}&${PARAM_PAGE}${page}&${PARAM_HPP}${DEFAULT_HPP}`;
    this.setState({isAjax : true});   
    fetch(searchQuery)
         .then( response => response.json())
         .then( result => {
          this.setSearchTopStories(result);
          this.setState({isAjax : false});  

         })
         .catch(err => { 
          alert('Somethig went wrong!!');
          this.setState({isAjax : false});
         });
  }

  onSearchChange(e){
    this.setState({query : e.target.value}); 
    
  }

  onSubmit(e){    
    const {query} = this.state;
    this.setState({searchKey : query});
    if(this.needsToSearchTopStories(query)){
      this.fetchSearchTopStories(query, DEFAULT_PAGE);
    }
    e.preventDefault();
  }

  componentDidMount(){
    const {query} = this.state;
    this.setState({searchKey : query});
    this.fetchSearchTopStories(query, DEFAULT_PAGE);
  }
  render() {   
    const { query, results, isAjax, searchKey} = this.state; 
    const page = (results && results[searchKey] && results[searchKey].page) || 0;
    const list = (results && results[searchKey] && results[searchKey].hits) || [];
    return (
      <div className="App">
       <div className="page">
        <div className="interactions">
        <Search isAjax={isAjax} value={query} onChange={this.onSearchChange} onSubmit={this.onSubmit} >
          Search
        </Search>
       </div>
        {<Table list={list}/> }  
        <div className="interactions">
          <Button onClick={ () => this.fetchSearchTopStories(searchKey, page +1) } isAjax={isAjax}>
           {isAjax === true ? "Loading...." : "More..."}
          </Button>
        </div>
      </div>
      </div>
    );
  }
}

//Search bomponent
const Search = ({value, onChange, onSubmit, isAjax, children }) => {    
    return (
      <form onSubmit={onSubmit}>           
        <input value={value} type="text" onChange={onChange}/>
        <button type="submit" disabled={isAjax === true ? "disabled" : ""}>{ children } </button>
      </form> 
      ); 
}

//Show list table component
const Table = ({list}) =>  {    
    return(
        <div className="table">
          {
           list.map( item =>           
                  <div key={item.objectID} className="table-row">
                    <span style={{width : '40%'}} ><a href={item.url}>{item.title}</a></span>
                    <span style={{width : '30%'}} >{item.author}</span>
                    <span style={{width : '15%'}} >{item.num_comments}</span>
                    <span style={{width : '15%'}} >{item.points}</span>
                  </div>            
           )
          }
        </div>
      );  
}

const Button = ({onClick, isAjax, children}) => {
 return (<button disabled={isAjax === true ? "disabled" : ""} onClick={onClick}>{children}</button>);
}
export default App;
export{
  Button,
  Search,
  Table
}
