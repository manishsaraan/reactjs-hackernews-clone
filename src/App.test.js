import React from 'react';
import ReactDOM from 'react-dom';
import renderer from 'react-test-renderer';
import { shallow } from 'enzyme';
import App from './App';
import {Search, Button, Table } from './App';


//test for search
describe('Search', () => {
  it('renders', () => {
  	const div = document.createElement('div');
  	ReactDOM.render(<Search>Search</Search>, div);
  });

  test('snapshots', () => {
    const component = renderer.create(<Search>Search</Search>);
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});

//test for button
describe('Button', () => {
  it('renders', () => {
  	const div = document.createElement('div');
  	ReactDOM.render(<Button>Give me more...</Button>, div);
  });

  test('snapshots', () => {
    const component = renderer.create(<Button>Give me more...</Button>);
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});

describe('Table', () => {
  const props = {
  	    list: [
				{ title: '1', author: '1', num_comments: 1, points: 2, objectID: 'y' },
				{ title: '2', author: '2', num_comments: 1, points: 2, objectID: 'z' },
			  ],
		}
  it('renders', () => {
    const div = document.createElement('div');
    ReactDOM.render(<Table {...props} />, div);
  });

  test('snapshots', () => {
    const component = renderer.create(<Table {...props}/>);
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('show two items of list', () => {
    const element = shallow(<Table {...props}/>);
    
    expect(element.find('.table-row').length).toBe(2);
  });
});

