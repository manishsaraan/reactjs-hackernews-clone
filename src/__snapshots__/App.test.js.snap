// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App snapshots 1`] = `
<div
  className="App"
>
  <div
    className="page"
  >
    <div
      className="interactions"
    >
      <form
        onSubmit={[Function]}
      >
        <input
          onChange={[Function]}
          type="text"
          value="redux"
        />
        <button
          disabled="disabled"
          type="submit"
        >
          Search
           
        </button>
      </form>
    </div>
    <div
      className="table"
    />
    <div
      className="interactions"
    >
      <button
        disabled="disabled"
        onClick={[Function]}
      >
        Loading....
      </button>
    </div>
  </div>
</div>
`;

exports[`Button snapshots 1`] = `
<button
  disabled=""
  onClick={undefined}
>
  Give me more...
</button>
`;

exports[`Search snapshots 1`] = `
<form
  onSubmit={undefined}
>
  <input
    onChange={undefined}
    type="text"
    value={undefined}
  />
  <button
    disabled=""
    type="submit"
  >
    Search
     
  </button>
</form>
`;

exports[`Table snapshots 1`] = `
<div
  className="table"
>
  <div
    className="table-row"
  >
    <span
      style={
        Object {
          "width": "40%",
        }
      }
    >
      <a
        href={undefined}
      >
        1
      </a>
    </span>
    <span
      style={
        Object {
          "width": "30%",
        }
      }
    >
      1
    </span>
    <span
      style={
        Object {
          "width": "15%",
        }
      }
    >
      1
    </span>
    <span
      style={
        Object {
          "width": "15%",
        }
      }
    >
      2
    </span>
  </div>
  <div
    className="table-row"
  >
    <span
      style={
        Object {
          "width": "40%",
        }
      }
    >
      <a
        href={undefined}
      >
        2
      </a>
    </span>
    <span
      style={
        Object {
          "width": "30%",
        }
      }
    >
      2
    </span>
    <span
      style={
        Object {
          "width": "15%",
        }
      }
    >
      1
    </span>
    <span
      style={
        Object {
          "width": "15%",
        }
      }
    >
      2
    </span>
  </div>
</div>
`;
